src/reconcile/rec_visibility.c

```diff
@@ -292,6 +292,36 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
         if (WT_TXNID_LT(max_txn, txnid))
             max_txn = txnid;
 
+        /*
+         * If we're seeing out-of-order updates, we need to pin these updates into the cache.
+         *
+         * We're also clearing the selected update since nothing newer than this out-of-order update
+         * is allowed to be written to the disk. If we find something older in the update list that
+         * can be written to the disk, that's ok and we should go ahead and select it.
+         *
+         * If the previous update is globally visible, we don't need to care about out-of-order ids
+         * or timestamps since the offending update will be removed by an obsolete check further
+         * down the line.
+         */
+        if (!F_ISSET(r, WT_REC_IN_MEMORY) && prev_upd != NULL &&
+          prev_upd->txnid != WT_TXN_ABORTED &&
+          !F_ISSET(
+            prev_upd, WT_UPDATE_HS | WT_UPDATE_RESTORED_FROM_DS | WT_UPDATE_RESTORED_FROM_HS))
```

Comment:
I don't think we need to check WT_UPDATE_RESTORED_FROM_DS | WT_UPDATE_RESTORED_FROM_HS, updates with these flags are guaranteed to be at the end of the chain.

Reply:
I was wondering this too. But I had a chat with @kommiharibabu before and he pointed out that in rollback to stable we write these updates to the head of the list (which I see in the code) which seems like it might affect it. I don't understand rollback to stable well so it's hard for me to say.

@kommiharibabu Does it seem fine to remove this?

Reply:
Even it is written to the head, there should be nothing on the update chain I believe @kommiharibabu. So effectively it is written to the end.

Reply:
Why should there be nothing in the chain? Do you just mean that they're aborted?

Reply:
Oh, I was wrong. We can call rollback to stable to trees still in memory.

Reply:
I was thinking we only call rollback to stable at start, in which case the tree is clean.

Reply:
For in-memory database, RTS doesn't restore any update from HS, because there is no HS. Yes, I agree that all the updates in the update chain are aborted whenever an update is restored from HS to the update chain.

Reply:
@kommiharibabu Is everything in the chain guaranteed to be aborted? What if some of those updates are stable?

If that is guaranteed, then I think we can remove this part of the check as @quchenhao suggested. The update may not be at the end of the list but everything after it will be aborted so it should effectively be the same.

Reply:
If there are some valid updates in the update list, we don't need to need to retrieve the update from history store.

Reply:
b8aa1ee